### PR TITLE
Feature/hpc intel

### LINKED ISSUE #27: clean up hpc-intel feature branch
The intel hpc containers are working but the different build scripts need to be unified with hpccm.

### PR DESCRIPTION
**Pull request by mmiesch**
This is the main outcome (other than the containers themselves and the benchmarking) of the work I've been doing over the past two months to get multi-node, multi-stage intel application containers running across nodes on HPC systems such as S4, Discover and AWS CloudFormation clusters.

The most important additions are the `intel19-impi-app.py` hpccm script and the modulefiles for S4 and Discover that resolve conflicts between the container environment and the host environment.  There are also changes to the intel build scripts to make use of the hpccm user options that have been added.

### CODE REVIEW

**Review by xinzhang8noaa — APPROVED** [review]

File: build_intel_dev_container.sh [hunk 1]
```diff
@@ -31,13 +36,33 @@ fi
...
+        ;;
+    "2")
+        hpccm --recipe ${CNAME}.py --userarg hpc="True" \
+                                             mellanox="True" \
+                                             pmi0="True" \
+                                             --format docker > Dockerfile.$CNAME
+        ;;
+    *)
+        echo "ERROR: unsupported HPC option"
+	exit 1
```
Comment (rickgrubin) [inline build_intel_dev_container.sh]: Nitpick: align case statement code

**Comment by rickgrubin** [comment]
General question: the Docker defn files call have a number of separate RUN steps to call 'apt-get' -- it's clear that this helps separate and make readable the various component sets installed in the container.

Are there any size or security concerns that might dictate squashing layers?

**Review by rickgrubin — APPROVED** [review]
Looks good

**Comment by mmiesch** [comment]
Thanks @rickgrubin for your review and question.  I don't know why the multiple run commands would affect size and security but maybe you know more than I do.  According to [Docker's guidelines](https://docs.docker.com/develop/develop-images/dockerfile_best-practices/), "In older versions of Docker, it was important that you minimized the number of layers in your images to ensure they were performant. " but this is apparently less of an issue now.  I thought grouping installs together was convenient for reusing the build cache on failed builds and for organizing the functionality of different layers but I haven't thought of the implications beyond that.  If you think it could be a significant improvement to combine some of these layers please create a ZenHub issue and I'll see if it makes a difference.   Probably beyond the scope of this PR though.

**Pull request merged**